src/DependencyInjection/Compiler/RegisterControllersPass.php

```diff
@@ -83,4 +86,18 @@ private function buildControllerDefinitionFromClass(string $class) : Definition
 
         return $definition;
     }
+
+
+    /**
+     * @param Definition $definition
+     * @return Definition
+     */
+    private function injectDependencies(Definition $definition) : Definition
```

Comment:
Injecting dependencies should be done in new CompilerPass. This only registers controllers.

Also, detection of setters should be done generally via Traits.

1. Detect if there is `DoctrineAwareTrait` in it
2. Then add `setDoctrineRegistry` method call. No need for `[new Reference('doctrine')]` argument there as there will work setter injection.

Same for `TwigAwareTrait`, `RouterAwareTrait` etc. All these can be in 1 CompilerPass.